src/software/ai/passing/pass_generator.h

```diff
@@ -7,135 +7,64 @@
- * involving the PassGenerator be done with executables built in "Release" in order to
- * maximize performance ("Release" can be 2-10x faster then "Debug").
  */
+template <class ZoneEnum>
 class PassGenerator
 {
-   public:
-    // Delete the default constructor, we want to force users to choose what
-    // pass quality they deem reasonable
-    PassGenerator() = delete;
-
-    // Delete the copy and assignment operators because this class really shouldn't
-    // need them and we don't want to risk doing anything nasty with the internal
-    // threading this class uses
-    PassGenerator& operator=(const PassGenerator&) = delete;
-    PassGenerator(const PassGenerator&)            = delete;
+    static_assert(std::is_enum<ZoneEnum>::value,
+                  "PassGenerator: ZoneEnum must be an zone id enum");
 
+   public:
     /**
-     * Create a PassGenerator with given parameters
+     * Creates a new PassGenerator with the given pitch_division.
      *
-     * @param world The world we're passing int
-     * @param passer_point The point we're passing from
-     * @param pass_type The type of pass we would like to perform.
-     *                  NOTE: this will _try_ to generate a pass of the type given,
-     *                  but it is not guaranteed, and can change during pass
-     *                  execution because of Passer/Receiver decisions
-     * @param passing_config The dynamic parameter config to use to tune the PassGenerator
-     * @param run_deterministically If true, this disables all threading so that
-     *                              the same sequence of calls to this function always
-     *                              returns the same values, no matter how much time
-     *                              passes between calls
-     */
-    explicit PassGenerator(const World& world, const Point& passer_point,
-                           const PassType& pass_type,
-                           std::shared_ptr<const PassingConfig> passing_config,
-                           const bool run_deterministically = false);
-
-    /**
-     * Updates the world
+     * The PassGenerator will use this pitch division to guide initial random samples
+     * in each zone after the pitch has been divided.
      *
-     * @param world
+     * @param pitch_division The pitch division to use when looking for passes
      */
-    void setWorld(World world);
+    explicit PassGenerator(
+        std::shared_ptr<const FieldPitchDivision<ZoneEnum>> pitch_division,
+        std::shared_ptr<const PassingConfig> passing_config);
 
     /**
-     * Updates the point that we are passing from
+     * Creates a PassEvaluation given a world and a field pitch division.
      *
-     * @param passer_point the point we are passing from
-     */
-    void setPasserPoint(Point passer_point);
-
-    /**
-     * Set the id of the robot performing the pass.
+     * NOTE: It is important that this function is able to run in 10ms.
```

Comment:
Would be good to elaborate on _why_ it needs to be fast, and why 10ms

Reply:
added a comment